pkg/api/pod/warnings.go

```diff
@@ -0,0 +1,240 @@
+limitations under the License.
+*/
+
+package pod
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/apis/core/pods"
+)
+
+func GetWarningsForPod(ctx context.Context, pod, oldPod *api.Pod) []string {
+	if pod == nil {
+		return nil
+	}
+
+	var (
+		oldSpec *api.PodSpec
+		oldMeta *metav1.ObjectMeta
+	)
+	if oldPod != nil {
+		oldSpec = &oldPod.Spec
+		oldMeta = &oldPod.ObjectMeta
+	}
+	return warningsForPodSpecAndMeta(nil, &pod.Spec, &pod.ObjectMeta, oldSpec, oldMeta)
+}
+
+func GetWarningsForPodTemplate(ctx context.Context, fieldPath *field.Path, podTemplate, oldPodTemplate *api.PodTemplateSpec) []string {
+	if podTemplate == nil {
+		return nil
+	}
+
+	var (
+		oldSpec *api.PodSpec
+		oldMeta *metav1.ObjectMeta
+	)
+	if oldPodTemplate != nil {
+		oldSpec = &oldPodTemplate.Spec
+		oldMeta = &oldPodTemplate.ObjectMeta
+	}
+	return warningsForPodSpecAndMeta(fieldPath, &podTemplate.Spec, &podTemplate.ObjectMeta, oldSpec, oldMeta)
+}
+
+var deprecatedNodeLabels = map[string]string{
+	`beta.kubernetes.io/arch`:                  `deprecated since v1.14; use "kubernetes.io/arch" instead`,
+	`beta.kubernetes.io/os`:                    `deprecated since v1.14; use "kubernetes.io/os" instead`,
+	`failure-domain.beta.kubernetes.io/region`: `deprecated since v1.17; use "topology.kubernetes.io/region" instead`,
+	`failure-domain.beta.kubernetes.io/zone`:   `deprecated since v1.17; use "topology.kubernetes.io/zone" instead`,
+	`beta.kubernetes.io/instance-type`:         `deprecated since v1.17; use "node.kubernetes.io/instance-type" instead`,
+}
+
+var deprecatedAnnotations = []struct {
+	key     string
+	prefix  string
+	message string
+}{
+	{
+		key:     `scheduler.alpha.kubernetes.io/critical-pod`,
+		message: `non-functional in v1.16+; use the "priorityClassName" field instead`,
+	},
+	{
+		key:     `seccomp.security.alpha.kubernetes.io/pod`,
+		prefix:  `container.seccomp.security.alpha.kubernetes.io/`,
+		message: `deprecated since v1.19; use the "seccompProfile" field instead`,
+	},
+	{
+		key:     `security.alpha.kubernetes.io/sysctls`,
+		message: `non-functional in v1.11+; use the "sysctls" field instead`,
+	},
+	{
+		key:     `security.alpha.kubernetes.io/unsafe-sysctls`,
+		message: `non-functional in v1.11+; use the "sysctls" field instead`,
+	},
+}
+
+func warningsForPodSpecAndMeta(fieldPath *field.Path, podSpec *api.PodSpec, meta *metav1.ObjectMeta, oldPodSpec *api.PodSpec, oldMeta *metav1.ObjectMeta) []string {
+	var warnings []string
+
+	// use of deprecated node labels in selectors/affinity/topology
+	for k := range podSpec.NodeSelector {
+		if msg, deprecated := deprecatedNodeLabels[k]; deprecated {
+			warnings = append(warnings, fieldPath.Child("spec", "nodeSelector").Key(k).String()+": "+msg)
```

Comment:
use a fmt.Sprintf istead of +

Reply:
really?

Reply:
yes!

Reply:
but... https://gist.github.com/dtjm/c6ebc86abe7515c988ec